cloudinit/sources/DataSourceAzure.py

```diff
@@ -596,25 +596,32 @@ def _poll_imds(self):
         return_val = None
 
         def exc_cb(msg, exception):
-            if isinstance(exception, UrlError) and exception.code == 404:
-                if self.imds_poll_counter == self.imds_logging_threshold:
-                    # Reducing the logging frequency as we are polling IMDS
-                    self.imds_logging_threshold *= 2
-                    LOG.debug("Call to IMDS with arguments %s failed "
-                              "with status code %s after %s retries",
-                              msg, exception.code, self.imds_poll_counter)
-                    LOG.debug("Backing off logging threshold for the same "
-                              "exception to %d", self.imds_logging_threshold)
-                self.imds_poll_counter += 1
-                return True
-
-            # If we get an exception while trying to call IMDS, we
-            # call DHCP and setup the ephemeral network to acquire the new IP.
-            LOG.debug("Call to IMDS with arguments %s failed  with "
-                      "status code %s", msg, exception.code)
-            report_diagnostic_event("polling IMDS failed with exception %s"
-                                    % exception.code)
-            return False
+            if isinstance(exception, UrlError):
+                if exception.code in (404, 410):
```

Comment:
can we add a @pyttest.mark.parameterize('exception_code', (404, 410)) decorator the the unit test https://github.com/canonical/cloud-init/blob/master/tests/unittests/test_datasource/test_azure.py#L1933
to exercise both 404 and 410 exception.code failure code paths.

If reworking the existing unit test to parametrize it is a bit too complex we could add a separate set of tests for this poll_imds method:

Reply:
I added the 410 error code to the unit test, The test now also makes more sense since it asserts that DHCP lease i not renewed for the case of 404 or 410.